docs/intermediate_representation/tensors.md

```diff
@@ -0,0 +1,154 @@
+# Tensor representation in the IR
+
+The ONNX IR offers the :class:`TensorProtocol` interface for usings different data structures as backing data for tensors. Besides the traditional ``onnx.TensorProto``, you can also use ``np.ndarray``, ``torch.Tensor``, and virtually anything else to represent tensors in the graph. This allows for  They can be accessed and serialized via the same `TensorProtocol` interface.
+
+## The ``TensorProtocol``
+
+:class:`ir.TensorProtocol` defines a read-only interface for representing tensors. A tensor class implementing the interface has attributes like ``name``, ``shape``, ``dtype``, ``size``, ``nbytes`` and ``metadata_props`` to describe basic properties of the tensor. Additionally, it implements two methods ``numpy()`` and ``__array__()`` which will produce equivalent numpy arrays from the backing data.
+
+When interacting with initializers, constant values and tensor attributes, it is best to assume ``TensorProtocol`` and only use ``isinstance`` to check for concrete classes when there is a need.
+
+## Tensor classes
+
+### TensorProto
+
+The ONNX spec defines [different ways](https://github.com/onnx/onnx/blob/d6f87121ba256ac6cc4d1da0463c300c278339d2/onnx/onnx.proto#L567-L654) for storing tensor data as a TensorProto protocol buffer message. The IR has corresponding classes for each of these data storage methods.
+
+For tensors whose data are stored in the ``raw_data`` field, we use the :class:`ir.TensorProtoTensor` as a wrapper around the proto to implement the ``ir.TensorProtocol`` interface. You can access ``shape``, ``dtype`` etc. as usual. A copy is incurred only when ``numpy()`` is called.
+
+```{eval-rst}
+.. exec_code::
+    import onnx
+    from onnxscript import ir
+
+    tensor_proto = onnx.helper.make_tensor(
+        "tensor", onnx.TensorProto.INT16, (3,), b'\x01\x00\x02\x00\x03\x00', raw=True
+    )
+    tensor = ir.TensorProtoTensor(tensor_proto)
+    print(tensor: ", tensor)  # TensorProtoTensor<INT16,[3]>(name='tensor')
+    print("shape: ", tensor.shape)  # ir.Shape([3])
+    print("dtype: ", tensor.dtype)  # ir.DataType.INT16
+    print(tensor.raw == tensor_proto)  # The raw field is the exact tensor_proto provided at initialization
+    print("tobytes: ", tensor.tobytes())  # b'\x01\x00\x02\x00\x03\x00'
+    print("numpy: ", tensor.numpy())  # array([1, 2, 3], dtype=int16)
+```
+
+When tensors store their data in a different data field like `int32_data`, use the corresponding class as a wrapper:
```

Comment:
Unclear. This looks like the same wrapper as the one above? But the description makes it sound like it is a different wrapper?

Reply:
Removed